src/art/art.c

```diff
@@ -1874,23 +2258,195 @@ static bool art_internal_validate_at(const art_node_t *node,
         return true;
     }
     art_internal_validate_t validator = {
         .reason = reason,
         .validate_cb = validate_cb,
+        .context = context,
         .depth = 0,
-        .current_key = {0},
+        .current_key = CROARING_ZERO_INITIALIZER,
     };
-    return art_internal_validate_at(art->root, validator);
+    for (art_typecode_t type = CROARING_ART_LEAF_TYPE;
+         type <= CROARING_ART_NODE256_TYPE; ++type) {
+        size_t capacity = art->capacities[type];
+        for (size_t i = 0; i < capacity; ++i) {
+            size_t first_free = art->first_free[type];
+            if (first_free > capacity) {
+                return art_validate_fail(&validator, "first_free > capacity");
+            }
+            size_t next_free = art_next_free(art, type, 0);
+            if (first_free != next_free) {
+                return art_validate_fail(&validator, "first_free != next_free");
+            }
+        }
+    }
+    return art_internal_validate_at(art, art->root, validator);
+}
+
+_Static_assert(alignof(art_leaf_t) == alignof(art_node4_t),
+               "Serialization assumes node type alignment is equal");
+_Static_assert(alignof(art_leaf_t) == alignof(art_node16_t),
+               "Serialization assumes node type alignment is equal");
+_Static_assert(alignof(art_leaf_t) == alignof(art_node48_t),
+               "Serialization assumes node type alignment is equal");
+_Static_assert(alignof(art_leaf_t) == alignof(art_node256_t),
+               "Serialization assumes node type alignment is equal");
+
+size_t art_size_in_bytes(const art_t *art) {
+    // Root.
+    size_t size = sizeof(art->root);
+    // Node counts.
+    size += sizeof(art->capacities);
+    // Alignment for leaves. The rest of the nodes are aligned the same way.
+    size +=
+        ((size + alignof(art_leaf_t) - 1) & ~(alignof(art_leaf_t) - 1)) - size;
+    size += art->capacities[CROARING_ART_LEAF_TYPE] * sizeof(art_leaf_t);
+    size += art->capacities[CROARING_ART_NODE4_TYPE] * sizeof(art_node4_t);
+    size += art->capacities[CROARING_ART_NODE16_TYPE] * sizeof(art_node16_t);
+    size += art->capacities[CROARING_ART_NODE48_TYPE] * sizeof(art_node48_t);
+    size += art->capacities[CROARING_ART_NODE256_TYPE] * sizeof(art_node256_t);
+    return size;
+}
+
+size_t art_serialize(const art_t *art, char *buf) {
+    if (buf == NULL) {
+        return 0;
+    }
+    const char *initial_buf = buf;
+
+    // Root.
+    memcpy(buf, &art->root, sizeof(art->root));
+    buf += sizeof(art->root);
+
+    // Node counts.
+    memcpy(buf, art->capacities, sizeof(art->capacities));
+    buf += sizeof(art->capacities);
+
+    if (art->capacities[CROARING_ART_LEAF_TYPE] > 0) {
+        size_t size =
+            art->capacities[CROARING_ART_LEAF_TYPE] * sizeof(art_leaf_t);
+        // Alignment for leaves. The rest of the nodes are aligned the same way.
+        buf =
+            CROARING_ART_ALIGN_RELATIVE(buf, initial_buf, alignof(art_leaf_t));
```

Comment:
We should zero any extra padding we add here.

Although I think realistically, there'll never be any extra alignment padding here, since we'll be at offset `8 + 6*sizeof(size_t)`, which will be evenly divisible by 8 on all but 16 bit platforms.